### ISSUE TITLE
torch.dot throws an error for input tensors of different dtypes

### ISSUE THREAD

**Issue by anjali411** [pytorch/pytorch#43064]
We should add type promotion for `torch.dot` and `[[WIP]torch.vdot](https://github.com/pytorch/pytorch/pull/43004)` to have a numpy like API.

```
>>> torch.dot(torch.tensor([1j]), torch.tensor([2]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dot : expected both vectors to have same dtype, but found ComplexFloat and Long
```

**Comment by rgommers** [comment]
Not quite sure about that. Had an interesting discussion recently about the commonality in type casting rules between the various array/tensor libraries. The consensus seemed to be that upcasting within the same "family" of dtypes (e.g. all integers, all floats, or all complex) but not allowing mixed int-float etc. casting was the best way to go.

An issue is that NumPy is very aggressive with upcasting when you mix types, which is typically not what you want for PyTorch:
```
In [20]: x = np.arange(3, dtype=int)                                          

In [21]: y = np.arange(3, dtype=np.float16)                                   

In [22]: (x + y).dtype                                                        
Out[22]: dtype('float64')
```

Mixing with complex dtypes will have similar issues.

**Comment by mruberry** [comment]
> Not quite sure about that. Had an interesting discussion recently about the commonality in type casting rules between the various array/tensor libraries. The consensus seemed to be that upcasting within the same "family" of dtypes (e.g. all integers, all floats, or all complex) but not allowing mixed int-float etc. casting was the best way to go.

Why?

> An issue is that NumPy is very aggressive with upcasting when you mix types, which is typically not what you want for PyTorch:

Our type promotion system is similar to JAX's and we bias towards float types, so long x fp16 -> fp16.

**Comment by rgommers** [comment]
> Our type promotion system is similar to JAX's and we bias towards float types, so long x fp16 -> fp16.

Yes I think that's the sensible behavior. I read the proposal here was to match NumPy (there, e.g., `long` x `float16` -> `float64`). Sounds like it's not. Okay, then please ignore my comment.
